pkg/await/extensions_ingress.go

```diff
@@ -0,0 +1,362 @@
+	if err != nil {
+		return errors.Wrapf(err,
+			"Could not create watcher for Endpoint objects associated with Ingress %q",
+			iia.config.currentInputs.GetName())
+	}
+	defer endpointWatcher.Stop()
+
+	return iia.await(ingressWatcher, endpointWatcher, time.After(10*time.Minute), make(chan struct{}))
+}
+
+func (iia *ingressInitAwaiter) initializeClients() error {
+	if iia.ingressClient == nil {
+		iia.ingressClient = &iia.config.clientForResource
+	}
+
+	if iia.endpointClient == nil {
+		client, err := client.FromGVK(iia.config.pool, iia.config.disco, schema.GroupVersionKind{
+			Group:   "",
+			Version: "v1",
+			Kind:    "Endpoints",
+		}, iia.config.currentInputs.GetNamespace())
+		if err != nil {
+			glog.V(3).Infof("Failed to initialize Endpoints client: %v", err)
+			return err
+		}
+
+		iia.endpointClient = &client
+	}
+
+	return nil
+}
+
+func (iia *ingressInitAwaiter) Read() error {
+	if err := iia.initializeClients(); err != nil {
+		glog.V(3).Infof("Failed to initialize client(s): %v", err)
+		return errors.Wrapf(err, "Failed to initialize client(s) required for Ingress awaiter")
+	}
+
+	// Get live versions of Ingress.
+	ingress, err := (*iia.ingressClient).Get(iia.config.currentInputs.GetName(), metav1.GetOptions{})
+	if err != nil {
+		// IMPORTANT: Do not wrap this error! If this is a 404, the provider need to know so that it
+		// can mark the deployment as having been deleted.
+		return err
+	}
+
+	// Get live version of Endpoints.
+	endpointList, err := (*iia.endpointClient).List(metav1.ListOptions{})
+	if err != nil {
+		glog.V(3).Infof("Failed to list endpoints needed for Ingress awaiter: %v", err)
+		endpointList = &unstructured.UnstructuredList{Items: []unstructured.Unstructured{}}
+	}
+
+	return iia.read(ingress, endpointList.(*unstructured.UnstructuredList))
+}
+
+func (iia *ingressInitAwaiter) read(
+	ingress *unstructured.Unstructured,
+	endpoints *unstructured.UnstructuredList,
+) error {
+	iia.processIngressEvent(watchAddedEvent(ingress))
+
+	var err error
+	settled := make(chan struct{})
+
+	glog.V(3).Infof("Processing endpoint list: %#v", endpoints)
+	err = endpoints.EachListItem(func(endpoint runtime.Object) error {
+		iia.processEndpointEvent(watchAddedEvent(endpoint.(*unstructured.Unstructured)), settled)
+		return nil
+	})
+	if err != nil {
+		glog.V(3).Infof("Error iterating over endpoint list for ingress %q: %v", ingress.GetName(), err)
+	}
+
+	iia.endpointsSettled = true
+	if iia.checkAndLogStatus() {
+		return nil
+	}
+
+	return &initializationError{
+		subErrors: iia.errorMessages(),
+		object:    ingress,
+	}
+}
+
+// await is a helper companion to `Await` designed to make it easy to test this module.
+func (iia *ingressInitAwaiter) await(
+	ingressWatcher, endpointWatcher watch.Interface, timeout <-chan time.Time,
+	settled chan struct{},
+) error {
+	iia.config.logStatus(diag.Info, "[1/3] Waiting on endpoints for each Ingress path")
```

Comment:
In the `Service` we chose to phrase this along the lines of "service must target pods". I think this is important to keep because many of our users do not know what endpoints are.